### PR TITLE
Add TransformerLayer, TransformerBlock, C3TR modules

### LINKED ISSUE #2329: Apply Transformer in the backbone
## 🚀 Feature

Transformer is popular in NLP, and now is also applied on CV. I added `C3TR` just by replacing the sequential `self.m` in `C3` with a Transformer block, which could reduce GFlOPs and make Yolo achieve a better result.

## Motivation

- Dosovitskiy et al. proposed [ViT](https://arxiv.org/pdf/2010.11929.pdf)
- Facebook applied [Transformer](https://ai.facebook.com/research/publications/end-to-end-object-detection-with-transformers) on object detection as an encoder
- So I thought Transformer could make yolo better

## Pitch

I add 3 classes in https://github.com/dingyiwei/yolov5/blob/Transformer/models/common.py :

```python
class TransformerLayer(nn.Module):
    def __init__(self, c, num_heads):
        super().__init__()

        self.ln1 = nn.LayerNorm(c)
        self.q = nn.Linear(c, c, bias=False)
        self.k = nn.Linear(c, c, bias=False)
        self.v = nn.Linear(c, c, bias=False)
        self.ma = nn.MultiheadAttention(embed_dim=c, num_heads=num_heads)
        self.ln2 = nn.LayerNorm(c)
        self.fc1 = nn.Linear(c, c, bias=False)
        self.fc2 = nn.Linear(c, c, bias=False)

    def forward(self, x):
        x_ = self.ln1(x)
        x = self.ma(self.q(x_), self.k(x_), self.v(x_))[0] + x
        x = self.ln2(x)
        x = self.fc2(self.fc1(x)) + x
        return x


class TransformerBlock(nn.Module):
    def __init__(self, c1, c2, num_heads, num_layers):
        super().__init__()

        self.conv = None
        if c1 != c2:
            self.conv = Conv(c1, c2)
        self.linear = nn.Linear(c2, c2)
        self.tr = nn.Sequential(*[TransformerLayer(c2, num_heads) for _ in range(num_layers)])
        self.c2 = c2

    def forward(self, x):
        if self.conv is not None:
            x = self.conv(x)
        b, _, w, h = x.shape
        p = x.flatten(2)
        p = p.unsqueeze(0)
        p = p.transpose(0, 3)
        p = p.squeeze(3)
        e = self.linear(p)
        x = p + e

        x = self.tr(x)
        x = x.unsqueeze(3)
        x = x.transpose(0, 3)
        x = x.reshape(b, self.c2, w, h)
        return x


class C3TR(C3):
    def __init__(self, c1, c2, n=1, shortcut=True, g=1, e=0.5):
        super().__init__(c1, c2, n, shortcut, g, e)
        c_ = int(c2 * e)
        self.m = TransformerBlock(c_, c_, 4, n)
```

And I just put it as the last part of the backbone instead of a `C3` block.

```
backbone:
  # [from, number, module, args]
  [[-1, 1, Focus, [64, 3]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, C3, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 9, C3, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, C3, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 1, SPP, [1024, [5, 9, 13]]],
   [-1, 3, C3TR, [1024, False]],  # 9    <---- here is my modifcation
  ]
```

I conducted experiments on 2 Nvidia GTX 1080Ti cards, where `depth_multiple` and `width_multiple` are the same as Yolov5s. Here are my experimental results with `img-size` 640. For convenience I named the method in this issue as Yolov5TRs.

|Model|Params|GFLOPs|
|---|---|---|
|Yolov5s|7266973|17.0|
|Yolov5TRs|7268765|16.8|

|Model|Dataset|TTA|mAP@.5|mAP@.5:.95|Speed (ms)|
|---|---|---|---|---|---|
|Yolov5s|coco (val)|N|0.558|0.365|4.4|
|Yolov5TRs|coco (val)|N|0.568|0.363|4.4|
|Yolov5s|coco (test-dev)|N|0.559|0.365|4.6|
|Yolov5TRs|coco (test-dev)|N|0.567|0.365|4.5|
|Yolov5s|coco (test-dev)|Y|0.568|0.378|12.0|
|Yolov5TRs|coco (test-dev)|Y|0.571|0.375|11.0|

We can see that Yolov5TRs get higher scores in mAP@0.5 with a faster speed. (I'm not sure why my results of Yolov5s are different from which shown in README. The model was downloaded from release v4.0) When `depth_multiple` and `width_multiple` are set to larger numbers, `C3TR` should be more lightweight than `C3`. Since I do not have so much time on it and my machine is not very strong, I did not run experiments on M, L and X. Maybe someone could conduct the future experiments:smile:

### PR DESCRIPTION
**Pull request by dingyiwei**
Transformer is popular in NLP, and now is also applied on CV. I added C3TR just by replacing the sequential self.m in C3 with a Transformer block, which could reduce GFlOPs and make Yolo achieve a better result.

Details can be seen in #2329 .

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Integration of Vision Transformer architecture into the YOLOv5 model.

### 📊 Key Changes
- 🦾 Added a new `TransformerLayer` class, with multi-head self-attention but without LayerNorm for performance.
- 🧠 Introduced `TransformerBlock` class, incorporating the Vision Transformer (ViT) architecture.
- 🧬 Created a new `C3TR` class as a variant of the `C3` module, using `TransformerBlock` to enhance object recognition capabilities.
- 🚀 Deployed an updated YOLOv5 architecture with transformer blocks in a new `yolov5s-transformer.yaml` config file.
- 👁‍🗨 Modified the `train.py` to handle potential compatibility issues between `nn.MultiheadAttention` and PyTorch's Distributed Data Parallel (DDP) training.

### 🎯 Purpose & Impact
- 🔄 The modifications aim to combine convolutional and transformer-based approaches for potential improvements in object detection performance.
- 🤖 By incorporating transformer architectures, the model may generalize better and identify objects with higher accuracy, benefiting users who need state-of-the-art detection models.
- 🌐 The DDP modification ensures that the new architecture works well with distributed training environments, maintaining efficient scaling for training on large datasets or across multiple GPUs.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@dingyiwei ah I see what you did there, you initialized C3TR from C3, I didn't notice that part before, very nice.

Ok I will start some experiments from this new PR branch!

**Comment by zhiqwang** [comment]
Hi @dingyiwei ,

This result is awesome! BTW, do you plan to upload the trained model weights : )

**Comment by dingyiwei** [comment]
@zhiqwang Oh I forgot to upload it.. A trained model is now on [GoogleDrive](https://drive.google.com/file/d/18dmCEjssnh1N90PWGnNYUsvfiQyDeJqZ/view?usp=sharing).

**Comment by dingyiwei** [comment]
Hi @glenn-jocher , the mistake mentioned in [#2329 (comment)](https://github.com/ultralytics/yolov5/issues/2329#issuecomment-790372849) is fixed in the last commit.

**Comment by glenn-jocher** [comment]
I think regardless of experiment results we should merge this for future experimentation. Will try to get this done soon! 
/rebase

**Comment by glenn-jocher** [comment]
@dingyiwei thanks for the updates! Is this ready to merge now?

**Comment by glenn-jocher** [comment]
I think we can simply find_unused_params:

```python
find_unused_params = False if not [type(layer) for layer in model.modules() if isinstance(layer, nn.MultiheadAttention)] else True
# replace with this?
find_unused_params = any([isinstance(layer, nn.MultiheadAttention) for layer in model.modules()]) 
```

**Comment by glenn-jocher** [comment]
EDIT: can remove brackets also:
```python
find_unused_params = any(isinstance(layer, nn.MultiheadAttention) for layer in model.modules())
```

**Comment by dingyiwei** [comment]
@glenn-jocher Thanks for the promotion! I think it is good to merge.

**Comment by glenn-jocher** [comment]
@dingyiwei PR is merged, thank you for your contributions!

**Comment by QY1994-0919** [comment]
> Transformer is popular in NLP, and now is also applied on CV. I added C3TR just by replacing the sequential self.m in C3 with a Transformer block, which could reduce GFlOPs and make Yolo achieve a better result.
> 
> Details can be seen in #2329 .

Hello!
     I loaded your modified C3T and the official C3TR into YOLOv5 respectively, and made a comparison on coco128, and found that the result of C3T is not as good as C3TR. What is the reason? And the model Summary is also different.

**Comment by QY1994-0919** [comment]
@dingyiwei Hello!
     I loaded your modified C3T and the official C3TR into YOLOv5 respectively, and made a comparison on coco128, and found that the result of C3T is not as good as C3TR. What is the reason? And the model Summary is also different.

**Comment by dingyiwei** [comment]
Hi @QY1994-0919 , I didn't validate models on coco128 since it is too small to show the performance of different methods or components.

**Comment by glenn-jocher** [comment]
@dingyiwei @QY1994-0919 yes COCO128 is a 'sanity-check' dataset that you can use to verify your environment and training pipeline. It's not really meant for long trainings or deployments, as it's far too small to generalize well or produce reliable metrics as it also doesn't have a seperate val set.

**Comment by glenn-jocher** [comment]
One good option for fast object detection prototyping is VOC, as it's small enough that you can fully train in < a day but large enough that results there should extrapolate to larger datasets like COCO.
